doc/dev/error-categorization.md

```diff
@@ -0,0 +1,111 @@
+
+If not it will most likely fit into an existing category. Categories are leaf based, so you cannot
+put an error into a node such as `Permanent errors` (see below).
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+- Permanent errors ("01000")
+  - Resource ("01100")
+  - Parsing ("01200")
+  - Installation ("01300")
+  - Logical ("01400")
+- Conflict ("02000")
+- Timeout ("03000")
+- Validation ("04000")
+  - Syntactic ("04100")
+  - Semantic ("04200")
+### Permanent errors ("01000")
+
+The main category `Permanent Errors` refer to such errors which cannot be fixed by retry 
+at all. `Permanent Errors` are subdivided into
+- Resource
+- Parsing
+- Installation
+- Logical
+#### Resource ("01100")
+
+`Resource Errors` are all kinds of errors which are either permission related, existence related
+or missing resources such as memory, RAM, etc.
+Examples are missing files, insufficient permissions or out of memory.
+This category forces users or applications to provide additional resources, create a missing file/ directory etc
+or provide the relevant permissions for Elektra to work.
+
+#### Parsing ("01200")
+
+`Parsing Errors` are reserved for parsers for xml/yaml/ini etc. and indicate errors which
+were investigated during parsing. Such errors show unexpected encounters like missing line endings,
+illegal characters or wrong encodings.
+Users or applications will have to sanitize the input and retry again.
+
+#### Installation ("01300")
+
+`Installation Errors` are those errors which are related to Elektra's installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
+get rid of this error. Applications will most likely not be able to handle such errors by themselves.
+
+#### Logical ("01400")
+
+`Logical Errors` indicate a bug in Elektra such as internal errors, assertion failures or errors
+which should not happen such as going into a `default` branch when you are assured that all cases
+are covered. Usually such errors come with a message to report such failures to Elektra's bugtracker.
+Applications cannot handle such errors themselves other than reporting it to Elektra and hope
+that the developers investigate and fix the issue.
+
+### Conflict ("02000")
+
+`Conflict Errors` are errors which indicate temporary problems which can be handled by handled by
+updating the current state. Examples are the need for calling `kdbGet` before `kdbSet`, or pulling changes
+from git before pushing them. These kind of errors are usually in resolver plugins when the state of the file
+has changed without the system knowing. Applications interacting with such an error usually know
+how to handle them once approached such as the `highlevel API`.
```

Comment:
> such as the `highlevel API`.

This is not correct. The high-level API could only deal so easily with the old error code 30, because that was  a very specific error and the high-level API provides very specific circumstances. Most applications can only **attempt** a merge, there is no guarantee there won't be merge conflicts. Please mention that.

Also, if we categorise everything that fits the above description as `C02000`, the high-level API won't be able to handle this error anymore. (Doesn't really matter, but we should mention it here in that case).